### PR TITLE
SONARIAC-1383 Bicep parsing shouldn't fail if a string literal starts with a comment

### PR DESCRIPTION
**Pull request by nils-werner-sonarsource**
I've gotten rid of the StringComplete class. IMO there was no need for this tree type as the StringLiteral also does the job. Due to this, I changed a lot of files. Sorry

### CODE REVIEW

File: iac-extensions/arm/src/test/resources/checks/SecureParameterDefaultValueCheck/secureParameter.bicep [hunk 1]
```diff
@@ -10,7 +10,7 @@ param nonCompliant1 string = 'my secret' // Noncompliant{{Remove the default val
 param nonCompliant2 object = {key:'my secret'} // Noncompliant{{Remove the default value from this secure object.}}
 //                           ^^^^^^^^^^^^^^^^^
 @secure()
-param nonCompliant3 string = '[newGuid()]' // Noncompliant
+param nonCompliant3 string = '[newGuid()]' // False Negative, TODO SONARIAC-1038
```
Comment (nils-werner-sonarsource) [inline iac-extensions/arm/src/test/resources/checks/SecureParameterDefaultValueCheck/secureParameter.bicep:13]: I would like to accept this FN, as this is a super special edge case.

File: iac-extensions/arm/src/main/java/org/sonar/iac/arm/visitors/ArmHighlightingVisitor.java [hunk 2]
```diff
@@ -85,7 +84,6 @@ protected void languageSpecificHighlighting() {
     registerTree(AmbientTypeReference.class, KEYWORD);
     registerTree(InterpolatedString.class, STRING);
     registerTree(MultilineString.class, STRING);
-    registerTree(StringLiteral.class, STRING);
```
Comment (nils-werner-sonarsource) [inline iac-extensions/arm/src/main/java/org/sonar/iac/arm/visitors/ArmHighlightingVisitor.java:88]: StringLiteral now implements InterpolatedString.

File: iac-extensions/arm/src/main/java/org/sonar/iac/arm/parser/bicep/TreeFactory.java [hunk 3]
```diff
@@ -242,8 +242,8 @@ public ModuleDeclaration moduleDeclaration(
     return new ModuleDeclarationImpl(decorators.or(emptyList()), keyword, name, type, equals, value);
   }
 
-  public StringComplete stringComplete(SyntaxToken openingApostrophe, SyntaxToken value, SyntaxToken closingApostrophe) {
-    return new StringCompleteImpl(openingApostrophe, value, closingApostrophe);
+  public StringLiteral stringComplete(SyntaxToken value) {
```
Comment (petertrr) [inline iac-extensions/arm/src/main/java/org/sonar/iac/arm/parser/bicep/TreeFactory.java:245]: Method could be renamed

File: iac-extensions/arm/src/main/java/org/sonar/iac/arm/parser/bicep/BicepLexicalGrammar.java [hunk 4]
```diff
@@ -156,6 +156,7 @@ private static void lexical(LexerlessGrammarBuilder b) {
 
     b.rule(IDENTIFIER_LITERAL).is(SPACING, b.regexp(BicepLexicalConstant.IDENTIFIER_LITERAL));
     b.rule(QUOTED_STRING_LITERAL).is(SPACING, b.regexp(BicepLexicalConstant.QUOTED_STRING_LITERAL_NO_QUOTES));
+    b.rule(REGULAR_STRING_LITERAL).is(SPACING, b.regexp("'(?:[^'\\\\]*+(?:\\\\[\\s\\S])?+)*+'"));
```
Comment (petertrr) [inline iac-extensions/arm/src/main/java/org/sonar/iac/arm/parser/bicep/BicepLexicalGrammar.java:159]: Can it be `"'" + BicepLexicalConstant.QUOTED_STRING_LITERAL_NO_QUOTES + "'"`?

File: iac-extensions/arm/src/test/resources/checks/SecureParameterDefaultValueCheck/secureParameter.bicep (hunk 1 above)
Comment (petertrr) [reply iac-extensions/arm/src/test/resources/checks/SecureParameterDefaultValueCheck/secureParameter.bicep:13]: Maybe also reflect in this comment or in the ticket, that FN is caused by an overlap in JSON and Bicep logic?

**Review by petertrr — APPROVED** [review]
LGTM, please see my suggestions

File: iac-extensions/arm/src/main/java/org/sonar/iac/arm/checks/ShortBackupRetentionCheck.java [hunk 5]
```diff
@@ -160,7 +160,7 @@ private static String propertyValueAsStringOrNull(@Nullable Tree tree) {
     return Optional.ofNullable(tree)
       .map(Property.class::cast)
       .map(Property::value)
-      .filter(expr -> expr.is(ArmTree.Kind.STRING_LITERAL, ArmTree.Kind.STRING_COMPLETE))
+      .filter(expr -> expr.is(ArmTree.Kind.STRING_LITERAL, ArmTree.Kind.STRING_LITERAL))
```
[the remark this answers is not in the document]
Comment (nils-werner-sonarsource) [reply iac-extensions/arm/src/main/java/org/sonar/iac/arm/checks/ShortBackupRetentionCheck.java:163]: Yeah, I agree. One should be enough. 🙈
